src/adagrams.js

```diff
@@ -1,15 +1,177 @@
+  for (let i = letters.length - 1; i >= 1; i--) {
+    let j = Math.floor(Math.random() * (i + 1));
+    let temp = letters[j];
+    letters[j] = letters[i];
+    letters[i] = temp;
+  };
+  return letters.slice(0, 10);
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
   // Implement this method for wave 2
+  let inputLetterFreq = {};
+  for (const letter of input) {
+    if (letter.toUpperCase() in inputLetterFreq) {
+      inputLetterFreq[letter.toUpperCase()] += 1;
+    } else {
+      inputLetterFreq[letter.toUpperCase()] = 1;
+    }
+  };
+
+  let lettersInHandFreq = {};
+  for (const letter of lettersInHand) {
+    if (letter.toUpperCase() in lettersInHandFreq) {
+      lettersInHandFreq[letter.toUpperCase()] += 1;
+    } else {
+      lettersInHandFreq[letter.toUpperCase()] = 1;
+    }
+  };
+
+  for (const letter of input) {
+    /* If the letter does not exist in the `lettersInHandFreq` or the `input` letter 
+    frequency is greater than `lettersInHand`'s frequency, return false. Otherwise,
+    return true. */
+    if (typeof(lettersInHandFreq[letter.toUpperCase()]) === 'undefined' ||
+    inputLetterFreq[letter.toUpperCase()] > lettersInHandFreq[letter
+    .toUpperCase()]) {
+      return false;
+    }
+  }
+  return true;
 };
 
 export const scoreWord = (word) => {
   // Implement this method for wave 3
+  const scoreChart = {
+    A: 1,
+    B: 3,
+    C: 3,
+    D: 2,
+    E: 1,
+    F: 4,
+    G: 2,
+    H: 4,
+    I: 1,
+    J: 8,
+    K: 5,
+    L: 1,
+    M: 3, 
+    N: 1,
+    O: 1,
+    P: 3,
+    Q: 10,
+    R: 1,
+    S: 1,
+    T: 1,
+    U: 1,
+    V: 4,
+    W: 4,
+    X: 8,
+    Y: 4,
+    Z: 10
+  };
+
+  let score = 0;
+  for (const letter of word) {
+    if (letter.toUpperCase() in scoreChart) {
+      score += scoreChart[letter.toUpperCase()];
+    }
+  }
+
+  if (word.length >= 7 && word.length <= 10) {
+    score += 8;
+  }
+  return score;
 };
 
 export const highestScoreFrom = (words) => {
   // Implement this method for wave 1
-};
+
+  // create an array that holds each word and its score, respectively
+  // for example if `words` is ['X', 'XX', 'XXX', 'XXXX'], `wordScores` should be...
+  // ["X", 8, "XX", 16, "XXX", 24, "XXXX", 32]
+
+  const wordScores = [];
+  for (let i = 0; i < words.length; i++) {
+    wordScores.push(words[i])
+    wordScores.push(scoreWord(words[i]))
+  }
+
+  // find the max score in `wordScores`
+  let maxScore = 0;
+  for (let i = wordScores.length - 1; i >= 1; i -= 2) {
+    if (wordScores[i] > maxScore) {
+      maxScore = wordScores[i];
+    }
+  }
+
+  // find the words with a score that equals `maxScore`
+  let tieScoreWords = [];
+  for (let i = wordScores.length - 1; i >= 1; i -= 2) {
+    if (wordScores[i] === maxScore) {
+      tieScoreWords.push(wordScores[i - 1]);
+    }
+  }
```

Comment:
The tiebreaking rules want us to return the first item we find with the highest score if there are 2 words with the same length and score. When we iterate from the back of the list, we're pushing words into `tieScoreWords` in the opposite order from the input list, meaning we won't return the word that the tests are expecting. 

If we flip around our for loop and iterate from the start of the list, the code works exactly as the tests are expecting.